### ISSUE TITLE
Support `type(y) == x` type guard pattern when `x` has a generic type with an upper bound

### ISSUE THREAD

**Issue by dinatamas** [microsoft/pyright#9227]
# Bug Report

There seems to be a type narrowing bug where a type equality check is ignored under certain conditions.

## Reproducible Example

The `convert_if_not_already` function takes a type `x` and a variable `y`.
If `y` is already exactly of type `x`, it is returned as-is.
Otherwise, `y` is converted to type `x` and returned.

**Behavior:** Received `Type "A" is not assignable to return type "TT@convert_if_not_already"` on `return y`

**Expected Behavior:** The `type(y) == x` check should be enough to tell that `y` is of the required type.

Interestingly, if the `isinstance(y, A)` line is deleted and the following lines dedented, there is no error.

```python
from typing import TypeVar, Type, reveal_type

class A:
    def __init__(self, x):
        self.x = x

class B(A):
    pass

TT = TypeVar("TT", bound="A")

def convert_if_not_already(x: Type[TT], y) -> TT:
    if isinstance(y, A):
        if type(y) == x:
            reveal_type(y)  # If isinstance present: Type of "y" is "A"
                            # If isinstance commented out: Type of "y" is "A*"
            return y
    return x(y)

a1 = A(10)
a2 = convert_if_not_already(B, a1)
assert isinstance(a2, B)

b1 = B(10)
b2 = convert_if_not_already(B, b1)
assert b2 is b1

```

# Environment

The issue can be reproduced with the latest pyright version:

1. `python3 -V`
   _Python 3.12.2_
2. `python3 -m venv venv`
3. `source venv/bin/activate`
4. `python3 -m pip install pyright`
   _Successfully installed nodeenv-1.9.1 pyright-1.1.384 typing-extensions-4.12.2_
5. `pyright example.py`
   _Type "A" is not assignable to type "TT@convert_if_not_already" (reportReturnType)_

**Comment by erictraut** [comment]
Pyright's type narrowing logic for the `type(y) == x` pattern is working as intended here, so I don't consider this a bug. The narrowing logic doesn't currently handle the case where `y` is a type variable and `x` is concrete.

I will need to think about this more, but it may be a a reasonable enhancement request to support this case. I noticed that mypy has some limited support for this case, specifically where the type variable has an upper bound that overlaps with the type of `x`. If there is no such overlap, mypy's narrowing logic evaluates an incorrect type. Mypy also doesn't properly handle the case where the class is generic. There are some complicated gotchas with this pattern that require some additional thought.

For my own benefit, here's a minimal code sample.

```python
class A:
    def __init__(self, x): ...


def func1[T: A](x: type[T], y: A) -> T:
    if type(y) == x:
        reveal_type(y)
        return y
    return x(y)
```

Note that the behavior you're seeing isn't dependent on an `isinstance` check.

I'll give this enhancement request more thought and get back to you.

**Comment by erictraut** [comment]
This enhancement will be included in the next release.

**Comment by dinatamas** [comment]
Wow, thank you @erictraut ! :)

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.386
